pulp_ansible/tests/unit/test_search.py

```diff
@@ -0,0 +1,157 @@
+        for repo_name in ["published1", "published2"]:
+            self.repos[repo_name] = AnsibleRepository(name=repo_name)
+            self.repos[repo_name].save()
+
+        # make a distribution for each repo?
+        for repo_name, repo in self.repos.items():
+            AnsibleDistribution.objects.create(name=repo_name, base_path=repo_name, repository=repo)
+
+        # make some tarballs
+        self.collections = {}
+        for i in range(0, 2):
+            namespace_name = "".join([random.choice(string.ascii_lowercase) for x in range(0, 5)])
+            for j in range(0, 3):
+                collection_name = "".join(
+                    [random.choice(string.ascii_lowercase) for x in range(0, 5)]
+                )
+                for v in range(1, 3):
+                    vstring = f"1.0.{v}"
+                    tarfn = make_cv_tarball(namespace_name, collection_name, vstring)
+                    self.collections[(namespace_name, collection_name, vstring)] = {
+                        "tags": ["tag" + namespace_name + collection_name + vstring],
+                        "tar": tarfn,
+                    }
+
+        # "import" the collections
+        specs = sorted(list(self.collections.keys()))
+        for ids, spec in enumerate(specs):
+            cdata = self.collections[spec]
+
+            this_repo_name = random.choice(list(self.repos.keys()))
+            this_repo = self.repos[this_repo_name]
+            self.collections[spec]["repo"] = this_repo
+            self.collections[spec]["repo_name"] = this_repo_name
+
+            # make an artifact
+            rawbin = open(cdata["tar"], "rb").read()
+            artifact = Artifact.objects.create(
+                sha224=hashlib.sha224(rawbin).hexdigest(),
+                sha256=hashlib.sha256(rawbin).hexdigest(),
+                sha384=hashlib.sha384(rawbin).hexdigest(),
+                sha512=hashlib.sha512(rawbin).hexdigest(),
+                size=os.path.getsize(tarfn),
+                file=SimpleUploadedFile(tarfn, rawbin),
+            )
+            artifact.save()
+
+            # make the collection
+            col, _ = Collection.objects.get_or_create(name=spec[0])
+            col.save()
+            self.collections[spec]["col"] = col
+
+            # make the collection version
+            cv = CollectionVersion(collection=col, namespace=spec[0], name=spec[1], version=spec[2])
+            cv.save()
+            self.collections[spec]["cv"] = cv
+
+            # add tags ...
+            #   ansible_collectionversion -> ansible_collectionversion_tags -> ansible_tags
+            for tag_name in cdata["tags"]:
+                this_tag, _ = Tag.objects.get_or_create(name=tag_name)
+                cv.tags.add(this_tag)
+                cv.save()
+
+            # trigger an update and rebuild of the search vector
+            #   THIS IS THE ONLY WAY THAT THE SEARCH VECTOR IS CREATED!
+            cv.is_highest = False
+            cv.save()
```

Comment:
There's a translation/many-to-many table called ansible_collectionversion_tags that holds a reference to the collectionversion id and the tag id. Adding tags to a collectionversion updates the ansible_collectionversion_tags and ansible_tag tables only.

Reply:
So is the `cv.save()` after adding the tag a no-op, or is updating the in memory model by reading from the database?

Reply:
I'm not sure, but I'll go check and update here once done.

Reply:
You are right ... cv.save() after cv.tags.add() is unnecessary.

Reply:
for is_highest ...

Due to constraint errors, I can't set is_highest=True on all the versions without first setting it back to False for previous versions. Rather than compute and toggle that "normally", I tried just setting it to False to see if it would trigger the update, and sure enough it does. I -believe- django's ORM is making the SQL update call even though nothing changes and letting the database sort it out. I also believe that since the trigger is set to run "BEFORE" update, it may not compute that the rows changed is zero and continue on to build the search_vector regardless.